### PR TITLE
[System]: Fix memory leak in AppleTls.

### PR DESCRIPTION
**Pull request by baulig**

### CODE REVIEW

**Comment by migueldeicaza** [comment]
This looks good to me.

File: mcs/class/System/Mono.AppleTls/AppleTlsContext.cs [hunk 1]
```diff
@@ -73,8 +74,8 @@ class AppleTlsContext : MobileTlsContext
 			: base (parent, serverMode, targetHost, enabledProtocols,
 				serverCertificate, clientCertificates, askForClientCert)
 		{
-			handle = GCHandle.Alloc (this);
-			connectionId = GCHandle.ToIntPtr (handle);
+			var target = new WeakReference (this);
+			handle = GCHandle.Alloc (target);
```
Comment (rolfbjarne) [inline mcs/class/System/Mono.AppleTls/AppleTlsContext.cs]: Isn't this the same as `GCHandle.Alloc (this, GCHandleType.Weak)`?

File: mcs/class/System/Mono.AppleTls/AppleTlsContext.cs [hunk 2]
```diff
@@ -73,8 +74,8 @@ class AppleTlsContext : MobileTlsContext
 			: base (parent, serverMode, targetHost, enabledProtocols,
 				serverCertificate, clientCertificates, askForClientCert)
 		{
-			handle = GCHandle.Alloc (this);
-			connectionId = GCHandle.ToIntPtr (handle);
+			var target = new WeakReference (this);
```
Comment (alexrp) [inline mcs/class/System/Mono.AppleTls/AppleTlsContext.cs]: Why not just allocate a handle with `GCHandleType.Weak` ?

File: mcs/class/System/Mono.AppleTls/AppleTlsContext.cs [hunk 3]
```diff
@@ -684,18 +685,23 @@ public SecTrust GetPeerTrust (bool requireTrust)
...
+
+				context = (AppleTlsContext) weakReference.Target;
+				if (context == null || context.disposed)
+					return SslStatus.ClosedAbort;
 
-			try {
 				return context.NativeReadCallback (data, ref dataLength);
 			} catch (Exception ex) {
-				if (context.lastException == null)
+				if (context != null && context.lastException == null)
```
Comment (spouliot) [inline mcs/class/System/Mono.AppleTls/AppleTlsContext.cs]: `context` should not be `null` here, that case is already handled above

File: mcs/class/System/Mono.AppleTls/AppleTlsContext.cs [hunk 4]
```diff
@@ -704,18 +710,23 @@ static SslStatus NativeReadCallback (IntPtr ptr, IntPtr data, ref IntPtr dataLen
...
+
+				context = (AppleTlsContext) weakReference.Target;
+				if (context == null || context.disposed)
+					return SslStatus.ClosedAbort;
 
-			try {
 				return context.NativeWriteCallback (data, ref dataLength);
 			} catch (Exception ex) {
-				if (context.lastException == null)
+				if (context != null && context.lastException == null)
```
Comment (spouliot) [inline mcs/class/System/Mono.AppleTls/AppleTlsContext.cs]: same

File: mcs/class/System/Mono.AppleTls/AppleTlsContext.cs (hunk 3 above)
Comment (migueldeicaza) [inline mcs/class/System/Mono.AppleTls/AppleTlsContext.cs]: Not if `GCHandle.FromIntPtr` raises an exception.   My understanding is that Martin added that after someone from the runtime team asked for that, the original patch did not include it:

https://gist.github.com/migueldeicaza/83edb2f9e80bdc9fb34d91a9068dc80d

**Comment by migueldeicaza** [comment]
Patch with the suggestion on Weak:

https://gist.github.com/migueldeicaza/099fe9fabcadd32bd242b536377c6ab9

**Comment by chamons** [comment]
@baulig Did you run my sample with the latest patch?

I just rebuild XM w\ both of them and I get:

https://gist.github.com/chamons/be1e1eeee30941c2b3be3b8d0f53f005

100% of the time.

**Comment by chamons** [comment]
It looks like we are never calling Free on the weak reference being allocated. Discussions in #runtime suggest that this is likely leaking, though on a much smaller scale than the original leak.

File: mcs/class/System/Mono.AppleTls/AppleTlsContext.cs [hunk 5]
```diff
@@ -73,8 +74,7 @@ class AppleTlsContext : MobileTlsContext
 			: base (parent, serverMode, targetHost, enabledProtocols,
 				serverCertificate, clientCertificates, askForClientCert)
 		{
-			handle = GCHandle.Alloc (this);
-			connectionId = GCHandle.ToIntPtr (handle);
+			handle = GCHandle.Alloc (this, GCHandleType.Weak);
```
Comment (alanmcgovern) [inline mcs/class/System/Mono.AppleTls/AppleTlsContext.cs]: Does this change mean that the `this` object is now eligible for collection while we're in native code performing reads/writes? If it *is* eligible then it sounds like we've broken our guarantee. If it's *not* eligible for GC then the GC handle seems unnecessary.

File: mcs/class/System/Mono.AppleTls/AppleTlsContext.cs [hunk 6]
```diff
@@ -684,18 +684,19 @@ public SecTrust GetPeerTrust (bool requireTrust)
...
+			try {
+				var weakHandle = GCHandle.FromIntPtr (ptr);
+				if (!weakHandle.IsAllocated)
+					return SslStatus.Internal;
 
-			var context = (AppleTlsContext) handle.Target;
-			if (context.disposed)
-				return SslStatus.ClosedAbort;
+				context = (AppleTlsContext) weakHandle.Target;
+				if (context == null || context.disposed)
```
Comment (alanmcgovern) [inline mcs/class/System/Mono.AppleTls/AppleTlsContext.cs]: `context` is always null. We never assign to it

File: mcs/class/System/Mono.AppleTls/AppleTlsContext.cs [hunk 7]
```diff
@@ -704,18 +705,18 @@ static SslStatus NativeReadCallback (IntPtr ptr, IntPtr data, ref IntPtr dataLen
...
+			AppleTlsContext context = null;
+			try {
+				var weakHandle = GCHandle.FromIntPtr (ptr);
+				if (!weakHandle.IsAllocated)
+					return SslStatus.Internal;
 
-			var context = (AppleTlsContext) handle.Target;
-			if (context.disposed)
-				return SslStatus.ClosedAbort;
+				if (context == null || context.disposed)
```
Comment (alanmcgovern) [inline mcs/class/System/Mono.AppleTls/AppleTlsContext.cs]: `context` is always null. We never assign to it here.

Comment (baulig) [inline mcs/class/System/Mono.AppleTls/AppleTlsContext.cs]: Oops.

**Comment by baulig** [comment]
Should be fixed now.

**Comment by baulig** [comment]
build

**Pull request merged**